File: android/BOINC/app/src/test/java/edu/berkeley/boinc/rpc/AppsParserTest.java
/*
 * This file is part of BOINC.
 * http://boinc.berkeley.edu
 * Copyright (C) 2021 University of California
 *
 * BOINC is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * BOINC is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with BOINC.  If not, see <http://www.gnu.org/licenses/>.
 */
package edu.berkeley.boinc.rpc;

import android.util.Log;
import android.util.Xml;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;
import org.xml.sax.ContentHandler;
import org.xml.sax.SAXException;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.powermock.api.mockito.PowerMockito.doThrow;
import static org.powermock.api.mockito.PowerMockito.mockStatic;

@RunWith(PowerMockRunner.class)
@PrepareForTest({Log.class, Xml.class})
public class AppsParserTest {
    private static final String USER_FRIENDLY_NAME = "User-friendly name";

    private AppsParser appsParser;
    private App expected;

    @Before
    public void setUp() {
        appsParser = new AppsParser();
        expected = new App();
    }

    @Test
    public void testParse_whenRpcStringIsNull_thenExpectEmptyList() {
        mockStatic(Xml.class);

        assertTrue(AppsParser.parse(null).isEmpty());
    }

    @Test
    public void testParse_whenSAXExceptionIsThrown_thenExpectEmptyList() throws Exception {
        mockStatic(Xml.class);
        mockStatic(Log.class);

        doThrow(new SAXException()).when(Xml.class, "parse", anyString(), any(ContentHandler.class));

        assertTrue(AppsParser.parse("").isEmpty());
    }

    @Test
    public void testParser_whenOnlyStartElementIsRun_thenExpectElementStarted() throws SAXException {
        appsParser.startElement(null, "", null, null);

        assertTrue(appsParser.mElementStarted);
    }

    @Test
    public void testParser_whenBothStartElementAndEndElementAreRun_thenExpectElementNotStarted() throws SAXException {
        appsParser.startElement(null, AppsParser.APP_TAG, null, null);
        appsParser.endElement(null, AppsParser.APP_TAG, null);

        assertFalse(appsParser.mElementStarted);
    }

    @Test
    public void testParser_whenLocalNameIsEmpty_thenExpectEmptyList() throws SAXException {
        appsParser.startElement(null, "", null, null);

        assertTrue(appsParser.getApps().isEmpty());
    }

    @Test
    public void testParser_whenXmlAppHasNoElements_thenExpectEmptyList()
            throws SAXException {
        appsParser.startElement(null, AppsParser.APP_TAG, null, null);
        appsParser.endElement(null, AppsParser.APP_TAG, null);

        assertTrue(appsParser.getApps().isEmpty());
    }

    @Test
    public void testParser_whenXmlAppHasOnlyName_thenExpectMatchingApp()
            throws SAXException {
        appsParser.startElement(null, AppsParser.APP_TAG, null, null);
        appsParser.startElement(null, RPCCommonTags.NAME, null, null);
        appsParser.characters("Name".toCharArray(), 0, 4);
        appsParser.endElement(null, RPCCommonTags.NAME, null);
        appsParser.endElement(null, AppsParser.APP_TAG, null);

        expected.setName("Name");

        assertEquals(expected, appsParser.getApps().get(0));
    }

    @Test
    public void testParser_whenXmlAppHasOnlyNameAndUserFriendlyName_thenExpectMatchingApp()
            throws SAXException {
        appsParser.startElement(null, AppsParser.APP_TAG, null, null);
        appsParser.startElement(null, RPCCommonTags.NAME, null, null);
        appsParser.characters("Name".toCharArray(), 0, 4);
        appsParser.endElement(null, RPCCommonTags.NAME, null);
        appsParser.startElement(null, RPCCommonTags.USER_FRIENDLY_NAME, null, null);
        appsParser.characters(USER_FRIENDLY_NAME.toCharArray(), 0, USER_FRIENDLY_NAME.length());
        appsParser.endElement(null, RPCCommonTags.USER_FRIENDLY_NAME, null);
        appsParser.endElement(null, AppsParser.APP_TAG, null);

        expected.setName("Name");
        expected.setUserFriendlyName(USER_FRIENDLY_NAME);

        assertEquals(Collections.singletonList(expected), appsParser.getApps());
    }

    @Test
    public void testParser_whenXmlAppHasNameUserFriendlyNameAndInvalidNonCpuIntensive_thenExpectAppWithoutNonCpuIntensive()
            throws SAXException {
        mockStatic(Log.class);

        appsParser.startElement(null, AppsParser.APP_TAG, null, null);
        appsParser.startElement(null, RPCCommonTags.NAME, null, null);
        appsParser.characters("Name".toCharArray(), 0, 4);
        appsParser.endElement(null, RPCCommonTags.NAME, null);
        appsParser.startElement(null, RPCCommonTags.USER_FRIENDLY_NAME, null, null);
        appsParser.characters(USER_FRIENDLY_NAME.toCharArray(), 0, USER_FRIENDLY_NAME.length());
        appsParser.endElement(null, RPCCommonTags.USER_FRIENDLY_NAME, null);
        appsParser.startElement(null, RPCCommonTags.NON_CPU_INTENSIVE, null, null);
        appsParser.characters("One".toCharArray(), 0, 3);
        appsParser.endElement(null, RPCCommonTags.NON_CPU_INTENSIVE, null);
        appsParser.endElement(null, AppsParser.APP_TAG, null);

        expected.setName("Name");
        expected.setUserFriendlyName(USER_FRIENDLY_NAME);

        assertEquals(Collections.singletonList(expected), appsParser.getApps());
    }

    @Test
    public void testParser_whenXmlAppHasNameUserFriendlyNameAndNonCpuIntensive_thenExpectMatchingApp()
            throws SAXException {
        appsParser.startElement(null, AppsParser.APP_TAG, null, null);
        appsParser.startElement(null, RPCCommonTags.NAME, null, null);
        appsParser.characters("Name".toCharArray(), 0, 4);
        appsParser.endElement(null, RPCCommonTags.NAME, null);
        appsParser.startElement(null, RPCCommonTags.USER_FRIENDLY_NAME, null, null);
        appsParser.characters(USER_FRIENDLY_NAME.toCharArray(), 0, USER_FRIENDLY_NAME.length());
        appsParser.endElement(null, RPCCommonTags.USER_FRIENDLY_NAME, null);
        appsParser.startElement(null, RPCCommonTags.NON_CPU_INTENSIVE, null, null);
        appsParser.characters("1".toCharArray(), 0, 1);
        appsParser.endElement(null, RPCCommonTags.NON_CPU_INTENSIVE, null);
        appsParser.endElement(null, AppsParser.APP_TAG, null);

        expected.setName("Name");
        expected.setUserFriendlyName(USER_FRIENDLY_NAME);
        expected.setNonCpuIntensive(1);

        assertEquals(Collections.singletonList(expected), appsParser.getApps());
    }

    @Test
    public void testParser_whenTwoXmlAppsHaveNameUserFriendlyNameAndNonCpuIntensive_thenExpectTwoMatchingApps()
            throws SAXException {
        appsParser.startElement(null, AppsParser.APP_TAG, null, null);
        appsParser.startElement(null, RPCCommonTags.NAME, null, null);
        appsParser.characters("Name 1".toCharArray(), 0, 6);
        appsParser.endElement(null, RPCCommonTags.NAME, null);
        appsParser.startElement(null, RPCCommonTags.USER_FRIENDLY_NAME, null, null);
        appsParser.characters((USER_FRIENDLY_NAME + " 1").toCharArray(), 0, (USER_FRIENDLY_NAME +
                                                                             " 1").length());
        appsParser.endElement(null, RPCCommonTags.USER_FRIENDLY_NAME, null);
        appsParser.startElement(null, RPCCommonTags.NON_CPU_INTENSIVE, null, null);
        appsParser.characters("1".toCharArray(), 0, 1);
        appsParser.endElement(null, RPCCommonTags.NON_CPU_INTENSIVE, null);
        appsParser.endElement(null, AppsParser.APP_TAG, null);

        appsParser.startElement(null, AppsParser.APP_TAG, null, null);
        appsParser.startElement(null, RPCCommonTags.NAME, null, null);
        appsParser.characters("Name 2".toCharArray(), 0, 6);
        appsParser.endElement(null, RPCCommonTags.NAME, null);
        appsParser.startElement(null, RPCCommonTags.USER_FRIENDLY_NAME, null, null);
        appsParser.characters((USER_FRIENDLY_NAME + " 2").toCharArray(), 0, (USER_FRIENDLY_NAME +
                                                                             " 2").length());
        appsParser.endElement(null, RPCCommonTags.USER_FRIENDLY_NAME, null);
        appsParser.startElement(null, RPCCommonTags.NON_CPU_INTENSIVE, null, null);
        appsParser.characters("1".toCharArray(), 0, 1);
        appsParser.endElement(null, RPCCommonTags.NON_CPU_INTENSIVE, null);
        appsParser.endElement(null, AppsParser.APP_TAG, null);

        expected.setName("Name 1");
        expected.setUserFriendlyName(USER_FRIENDLY_NAME + " 1");
        expected.setNonCpuIntensive(1);

        List<App> apps = Arrays.asList(expected, new App("Name 2",
                                                         USER_FRIENDLY_NAME + " 2",
                                                         1));

        assertEquals(apps, appsParser.getApps());
    }
}

File: android/BOINC/app/src/test/java/edu/berkeley/boinc/rpc/AppsParserTest.kt
/*
 * This file is part of BOINC.
 * http://boinc.berkeley.edu
 * Copyright (C) 2021 University of California
 *
 * BOINC is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * BOINC is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with BOINC.  If not, see <http://www.gnu.org/licenses/>.
 */
package edu.berkeley.boinc.rpc

import android.util.Log
import android.util.Xml
import edu.berkeley.boinc.rpc.AppsParser.Companion.parse
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers
import org.powermock.api.mockito.PowerMockito
import org.powermock.core.classloader.annotations.PrepareForTest
import org.powermock.modules.junit4.PowerMockRunner
import org.xml.sax.ContentHandler
import org.xml.sax.SAXException

@RunWith(PowerMockRunner::class)
@PrepareForTest(Log::class, Xml::class)
class AppsParserTest {
    private lateinit var appsParser: AppsParser
    private lateinit var expected: App
    @Before
    fun setUp() {
        appsParser = AppsParser()
        expected = App()
    }

    @Test
    fun `When Rpc string is null then expect empty list`() {
        PowerMockito.mockStatic(Xml::class.java)
        Assert.assertTrue(parse(null).isEmpty())
    }

    @Test
    @Throws(Exception::class)
    fun `When SAXException is thrown then expect empty list`() {
        PowerMockito.mockStatic(Xml::class.java)
        PowerMockito.mockStatic(Log::class.java)
        PowerMockito.doThrow(SAXException()).`when`(
            Xml::class.java, "parse", ArgumentMatchers.anyString(), ArgumentMatchers.any(
                ContentHandler::class.java
            )
        )
        Assert.assertTrue(parse("").isEmpty())
    }

    @Test
    @Throws(SAXException::class)
    fun `When only start element is run then expect element started`() {
        appsParser.startElement(null, "", null, null)
        Assert.assertTrue(appsParser.mElementStarted)
    }

    @Test
    @Throws(SAXException::class)
    fun `When both start element and end element are run then expect element not started`() {
        appsParser.startElement(null, AppsParser.APP_TAG, null, null)
        appsParser.endElement(null, AppsParser.APP_TAG, null)
        Assert.assertFalse(appsParser.mElementStarted)
    }

    @Test
    @Throws(SAXException::class)
    fun `When 'localName' is empty then expect empty list`() {
        appsParser.startElement(null, "", null, null)
        Assert.assertTrue(appsParser.apps.isEmpty())
    }

    @Test
    @Throws(SAXException::class)
    fun `When Xml App has no elements then expect empty list`() {
        appsParser.startElement(null, AppsParser.APP_TAG, null, null)
        appsParser.endElement(null, AppsParser.APP_TAG, null)
        Assert.assertTrue(appsParser.apps.isEmpty())
    }

    @Test
    @Throws(SAXException::class)
    fun `When Xml App has only name then expect matching App`() {
        appsParser.startElement(null, AppsParser.APP_TAG, null, null)
        appsParser.startElement(null, NAME, null, null)
        appsParser.characters("Name".toCharArray(), 0, 4)
        appsParser.endElement(null, NAME, null)
        appsParser.endElement(null, AppsParser.APP_TAG, null)
        expected.name = "Name"
        Assert.assertEquals(expected, appsParser.apps[0])
    }

    @Test
    @Throws(SAXException::class)
    fun `When Xml App has only name and user_friendly_name then expect matching App`() {
        appsParser.startElement(null, AppsParser.APP_TAG, null, null)
        appsParser.startElement(null, NAME, null, null)
        appsParser.characters("Name".toCharArray(), 0, 4)
        appsParser.endElement(null, NAME, null)
        appsParser.startElement(null, edu.berkeley.boinc.rpc.USER_FRIENDLY_NAME, null, null)
        appsParser.characters(USER_FRIENDLY_NAME.toCharArray(), 0, USER_FRIENDLY_NAME.length)
        appsParser.endElement(null, edu.berkeley.boinc.rpc.USER_FRIENDLY_NAME, null)
        appsParser.endElement(null, AppsParser.APP_TAG, null)
        expected.name = "Name"
        expected.userFriendlyName = USER_FRIENDLY_NAME
        Assert.assertEquals(listOf(expected), appsParser.apps)
    }

    @Test
    @Throws(SAXException::class)
    fun `When Xml App has name, user_friendly_name and invalid non_cpu_intensive then expect App without non_cpu_intensive`() {
        PowerMockito.mockStatic(Log::class.java)
        appsParser.startElement(null, AppsParser.APP_TAG, null, null)
        appsParser.startElement(null, NAME, null, null)
        appsParser.characters("Name".toCharArray(), 0, 4)
        appsParser.endElement(null, NAME, null)
        appsParser.startElement(null, edu.berkeley.boinc.rpc.USER_FRIENDLY_NAME, null, null)
        appsParser.characters(USER_FRIENDLY_NAME.toCharArray(), 0, USER_FRIENDLY_NAME.length)
        appsParser.endElement(null, edu.berkeley.boinc.rpc.USER_FRIENDLY_NAME, null)
        appsParser.startElement(null, NON_CPU_INTENSIVE, null, null)
        appsParser.characters("One".toCharArray(), 0, 3)
        appsParser.endElement(null, NON_CPU_INTENSIVE, null)
        appsParser.endElement(null, AppsParser.APP_TAG, null)
        expected.name = "Name"
        expected.userFriendlyName = USER_FRIENDLY_NAME
        Assert.assertEquals(listOf(expected), appsParser.apps)
    }

    @Test
    @Throws(SAXException::class)
    fun `When Xml App has name, user_friendly_name and non_cpu_intensive then expect matching App`() {
        appsParser.startElement(null, AppsParser.APP_TAG, null, null)
        appsParser.startElement(null, NAME, null, null)
        appsParser.characters("Name".toCharArray(), 0, 4)
        appsParser.endElement(null, NAME, null)
        appsParser.startElement(null, edu.berkeley.boinc.rpc.USER_FRIENDLY_NAME, null, null)
        appsParser.characters(USER_FRIENDLY_NAME.toCharArray(), 0, USER_FRIENDLY_NAME.length)
        appsParser.endElement(null, edu.berkeley.boinc.rpc.USER_FRIENDLY_NAME, null)
        appsParser.startElement(null, NON_CPU_INTENSIVE, null, null)
        appsParser.characters("1".toCharArray(), 0, 1)
        appsParser.endElement(null, NON_CPU_INTENSIVE, null)
        appsParser.endElement(null, AppsParser.APP_TAG, null)
        expected.name = "Name"
        expected.userFriendlyName = USER_FRIENDLY_NAME
        expected.nonCpuIntensive = 1
        Assert.assertEquals(listOf(expected), appsParser.apps)
    }

    @Test
    @Throws(SAXException::class)
    fun `When two Xml Apps have name, user_friendly_name and non_cpu_intensive then expect two matching Apps`() {
        appsParser.startElement(null, AppsParser.APP_TAG, null, null)
        appsParser.startElement(null, NAME, null, null)
        appsParser.characters("Name 1".toCharArray(), 0, 6)
        appsParser.endElement(null, NAME, null)
        appsParser.startElement(null, edu.berkeley.boinc.rpc.USER_FRIENDLY_NAME, null, null)
        appsParser.characters(
            ("$USER_FRIENDLY_NAME 1").toCharArray(), 0, (USER_FRIENDLY_NAME +
                    " 1").length
        )
        appsParser.endElement(null, edu.berkeley.boinc.rpc.USER_FRIENDLY_NAME, null)
        appsParser.startElement(null, NON_CPU_INTENSIVE, null, null)
        appsParser.characters("1".toCharArray(), 0, 1)
        appsParser.endElement(null, NON_CPU_INTENSIVE, null)
        appsParser.endElement(null, AppsParser.APP_TAG, null)
        appsParser.startElement(null, AppsParser.APP_TAG, null, null)
        appsParser.startElement(null, NAME, null, null)
        appsParser.characters("Name 2".toCharArray(), 0, 6)
        appsParser.endElement(null, NAME, null)
        appsParser.startElement(null, edu.berkeley.boinc.rpc.USER_FRIENDLY_NAME, null, null)
        appsParser.characters(
            ("$USER_FRIENDLY_NAME 2").toCharArray(), 0, (USER_FRIENDLY_NAME +
                    " 2").length
        )
        appsParser.endElement(null, edu.berkeley.boinc.rpc.USER_FRIENDLY_NAME, null)
        appsParser.startElement(null, NON_CPU_INTENSIVE, null, null)
        appsParser.characters("1".toCharArray(), 0, 1)
        appsParser.endElement(null, NON_CPU_INTENSIVE, null)
        appsParser.endElement(null, AppsParser.APP_TAG, null)
        expected.name = "Name 1"
        expected.userFriendlyName = "$USER_FRIENDLY_NAME 1"
        expected.nonCpuIntensive = 1
        val apps: List<App?> = listOf(
            expected, App(
                "Name 2",
                "$USER_FRIENDLY_NAME 2",
                1
            )
        )
        Assert.assertEquals(apps, appsParser.apps)
    }

    companion object {
        private const val USER_FRIENDLY_NAME = "User-friendly name"
    }
}
